File: frontend/src/types/session-events.ts
/**
 * Typed live session event contracts (PROD-012).
 *
 * Matches backend app/schemas/session_events.py exactly.
 * The workspace UI renders structured state changes from these events
 * instead of parsing raw tutor text.
 */

// ---------------------------------------------------------------------------
// Event type enum
// ---------------------------------------------------------------------------

export type SessionEventType =
  | 'session.started'
  | 'session.resumed'
  | 'session.completed'
  | 'session.brief'
  | 'tutor.message.delta'
  | 'tutor.message.completed'
  | 'session.objective.updated'
  | 'session.objective.completed'
  | 'session.mastery.updated'
  | 'artifact.started'
  | 'artifact.updated'
  | 'artifact.completed'
  | 'checkpoint.requested'
  | 'checkpoint.response.received'
  | 'source.citation.available'
  | 'warning.model_rerouted'
  | 'warning.rate_limited';

// ---------------------------------------------------------------------------
// Payload shapes
// ---------------------------------------------------------------------------

export interface ObjectiveSnapshot {
  objective_id: string;
  title: string;
  description?: string | null;
  primary_concepts: string[];
  support_concepts: string[];
  prereq_concepts: string[];
  step_count: number;
  status: 'pending' | 'active' | 'completed' | 'skipped';
  progress_pct: number;
}

export interface SessionBriefPayload {
  notebook_id: string;
  session_id: string;
  mode: string;
  scope_type: string;
  resource_count: number;
  objectives_count: number;
  objectives: ObjectiveSnapshot[];
  mastery_snapshot: Record<string, number>;
  weak_concepts: string[];
  session_overview?: string | null;
}

export interface StructuredContentBlock {
  block_type: 'text' | 'concept_card' | 'quiz_card' | 'checkpoint' | 'latex' | 'diagram' | 'code';
  content: string;
  metadata: Record<string, unknown>;
}

export interface TutorMessageDeltaPayload {
  turn_id: string;
  delta: string;
  content_type: 'text' | 'markdown' | 'latex' | 'concept_card' | 'quiz_card';
}

export interface TutorMessageCompletedPayload {
  turn_id: string;
  response: string;
  tutor_question?: string | null;
  content_type: string;
  current_step?: string | null;
  current_step_index: number;
  objective_id?: string | null;
  objective_title?: string | null;
  step_transition?: string | null;
  focus_concepts: string[];
  pedagogical_action?: string | null;
  structured_content?: StructuredContentBlock[] | null;
}

export interface ObjectiveUpdatedPayload {
  objective_id: string;
  title: string;
  status: 'pending' | 'active' | 'completed' | 'skipped';
  progress_pct: number;
  attempts: number;
  correct: number;
}

export interface MasteryUpdatedPayload {
  concept_id: string;
  previous_score: number;
  new_score: number;
  delta: number;
}

export interface ArtifactEventPayload {
  artifact_id: string;
  artifact_type: 'notes' | 'flashcards' | 'quiz' | 'revision_plan' | 'concept_card';
  status: 'generating' | 'ready' | 'error';
  payload_json?: Record<string, unknown> | null;
  source_session_ids: string[];
  source_resource_ids: string[];
  error_message?: string | null;
}

export interface CheckpointRequestedPayload {
  checkpoint_id: string;
  checkpoint_type: 'understanding' | 'recall' | 'application';
  question: string;
  concept_id?: string | null;
  options: string[];
  allow_freeform: boolean;
}

export interface CheckpointResponsePayload {
  checkpoint_id: string;
  response: string;
  is_correct?: boolean | null;
  score?: number | null;
  feedback?: string | null;
}

export interface SourceCitationPayload {
  citation_id: string;
  resource_id: string;
  resource_name?: string | null;
  chunk_ids: string[];
  page_numbers: number[];
  snippet?: string | null;
}

export interface WarningPayload {
  warning_type: string;
  message: string;
  selected_model_id?: string | null;
  routed_model_id?: string | null;
  reroute_reason?: string | null;
}

export interface SessionCompletedPayload {
  session_id: string;
  summary_text?: string | null;
  concepts_strong: string[];
  concepts_developing: string[];
  concepts_to_revisit: string[];
  objectives: ObjectiveSnapshot[];
  mastery_snapshot: Record<string, number>;
  turn_count: number;
  recommended_next?: string | null;
}

// ---------------------------------------------------------------------------
// Study map snapshot (returned in turn responses)
// ---------------------------------------------------------------------------

export interface StudyMapStepSnapshot {
  type: string;
  goal: string;
  status: 'completed' | 'active' | 'upcoming' | 'skipped';
}

export interface StudyMapObjectiveSnapshot {
  objective_id: string;
  title: string;
  status: 'completed' | 'active' | 'pending';
  primary_concepts: string[];
  support_concepts: string[];
  prereq_concepts: string[];
  steps: StudyMapStepSnapshot[];
  progress: Record<string, number>;
}

export interface StudyMapSnapshot {
  current_objective_index: number;
  current_step_index: number;
  total_objectives: number;
  ad_hoc_count: number;
  max_ad_hoc: number;
  last_decision?: string | null;
  last_transition?: string | null;
  last_ad_hoc_type?: string | null;
  session_complete: boolean;
  objectives: StudyMapObjectiveSnapshot[];
}

// ---------------------------------------------------------------------------
// Event envelope
// ---------------------------------------------------------------------------

export interface SessionEvent {
  event_type: SessionEventType;
  timestamp: string;
  session_id: string;
  notebook_id?: string | null;
  payload: Record<string, unknown>;
  is_persistent: boolean;
}

// ---------------------------------------------------------------------------
// Type-safe event helpers
// ---------------------------------------------------------------------------

export type SessionEventPayloadMap = {
  'session.started': SessionBriefPayload;
  'session.resumed': SessionBriefPayload;
  'session.completed': SessionCompletedPayload;
  'session.brief': SessionBriefPayload;
  'tutor.message.delta': TutorMessageDeltaPayload;
  'tutor.message.completed': TutorMessageCompletedPayload;
  'session.objective.updated': ObjectiveUpdatedPayload;
  'session.objective.completed': ObjectiveUpdatedPayload;
  'session.mastery.updated': MasteryUpdatedPayload;
  'artifact.started': ArtifactEventPayload;
  'artifact.updated': ArtifactEventPayload;
  'artifact.completed': ArtifactEventPayload;
  'checkpoint.requested': CheckpointRequestedPayload;
  'checkpoint.response.received': CheckpointResponsePayload;
  'source.citation.available': SourceCitationPayload;
  'warning.model_rerouted': WarningPayload;
  'warning.rate_limited': WarningPayload;
};

export function parseSessionEvent<T extends SessionEventType>(
  event: SessionEvent,
  expectedType: T,
): SessionEventPayloadMap[T] | null {
  if (event.event_type !== expectedType) return null;
  return event.payload as unknown as SessionEventPayloadMap[T];
}

File: backend/app/schemas/session_events.py
"""Typed live session event contracts (PROD-012).

Canonical event families for the study workspace. Frontend and backend
share these shapes so the workspace UI can render structured state
changes instead of parsing raw tutor text.

Events are grouped into families:
  - session.*      – session lifecycle
  - tutor.*        – tutor message streaming
  - objective.*    – learning objective updates
  - mastery.*      – concept mastery changes
  - artifact.*     – generated artifact lifecycle
  - checkpoint.*   – understanding checkpoints
  - source.*       – citation / source metadata
  - warning.*      – operational warnings
"""

from __future__ import annotations

from datetime import datetime
from enum import Enum
from typing import Any, Dict, List, Optional
from uuid import UUID

from pydantic import BaseModel, Field


# ---------------------------------------------------------------------------
# Event type enum
# ---------------------------------------------------------------------------


class SessionEventType(str, Enum):
    # Session lifecycle
    SESSION_STARTED = "session.started"
    SESSION_RESUMED = "session.resumed"
    SESSION_COMPLETED = "session.completed"
    SESSION_BRIEF = "session.brief"

    # Tutor message
    TUTOR_MESSAGE_DELTA = "tutor.message.delta"
    TUTOR_MESSAGE_COMPLETED = "tutor.message.completed"

    # Objectives
    OBJECTIVE_UPDATED = "session.objective.updated"
    OBJECTIVE_COMPLETED = "session.objective.completed"

    # Mastery
    MASTERY_UPDATED = "session.mastery.updated"

    # Artifacts
    ARTIFACT_STARTED = "artifact.started"
    ARTIFACT_UPDATED = "artifact.updated"
    ARTIFACT_COMPLETED = "artifact.completed"

    # Checkpoints
    CHECKPOINT_REQUESTED = "checkpoint.requested"
    CHECKPOINT_RESPONSE_RECEIVED = "checkpoint.response.received"

    # Sources / citations
    SOURCE_CITATION_AVAILABLE = "source.citation.available"

    # Warnings
    WARNING_MODEL_REROUTED = "warning.model_rerouted"
    WARNING_RATE_LIMITED = "warning.rate_limited"


# ---------------------------------------------------------------------------
# Payload shapes
# ---------------------------------------------------------------------------


class SessionBriefPayload(BaseModel):
    """Sent at session start with context summary."""

    notebook_id: UUID
    session_id: UUID
    mode: str
    scope_type: str = "notebook"
    resource_count: int = 0
    objectives_count: int = 0
    objectives: List[ObjectiveSnapshot] = Field(default_factory=list)
    mastery_snapshot: Dict[str, float] = Field(default_factory=dict)
    weak_concepts: List[str] = Field(default_factory=list)
    session_overview: Optional[str] = None


class ObjectiveSnapshot(BaseModel):
    """Snapshot of a learning objective."""

    objective_id: str
    title: str
    description: Optional[str] = None
    primary_concepts: List[str] = Field(default_factory=list)
    support_concepts: List[str] = Field(default_factory=list)
    prereq_concepts: List[str] = Field(default_factory=list)
    step_count: int = 0
    status: str = "pending"  # pending | active | completed | skipped
    progress_pct: float = 0.0


class TutorMessageDeltaPayload(BaseModel):
    """Incremental tutor text chunk for streaming."""

    turn_id: str
    delta: str
    content_type: str = "text"  # text | markdown | latex | concept_card | quiz_card


class TutorMessageCompletedPayload(BaseModel):
    """Final tutor message with structured metadata."""

    turn_id: str
    response: str
    tutor_question: Optional[str] = None
    content_type: str = "text"
    current_step: Optional[str] = None
    current_step_index: int = 0
    objective_id: Optional[str] = None
    objective_title: Optional[str] = None
    step_transition: Optional[str] = None
    focus_concepts: List[str] = Field(default_factory=list)
    pedagogical_action: Optional[str] = None
    structured_content: Optional[List[StructuredContentBlock]] = None


class StructuredContentBlock(BaseModel):
    """A block of structured content within a tutor message."""

    block_type: (
        str  # text | concept_card | quiz_card | checkpoint | latex | diagram | code
    )
    content: str = ""
    metadata: Dict[str, Any] = Field(default_factory=dict)


class ObjectiveUpdatedPayload(BaseModel):
    """Objective progress changed."""

    objective_id: str
    title: str
    status: str  # pending | active | completed | skipped
    progress_pct: float = 0.0
    attempts: int = 0
    correct: int = 0


class MasteryUpdatedPayload(BaseModel):
    """Concept mastery changed."""

    concept_id: str
    previous_score: float = 0.0
    new_score: float = 0.0
    delta: float = 0.0


class ArtifactEventPayload(BaseModel):
    """Artifact lifecycle event."""

    artifact_id: str
    artifact_type: str  # notes | flashcards | quiz | revision_plan | concept_card
    status: str  # generating | ready | error
    payload_json: Optional[Dict[str, Any]] = None
    source_session_ids: List[str] = Field(default_factory=list)
    source_resource_ids: List[str] = Field(default_factory=list)
    error_message: Optional[str] = None


class CheckpointRequestedPayload(BaseModel):
    """Tutor requests understanding check from student."""

    checkpoint_id: str
    checkpoint_type: str = "understanding"  # understanding | recall | application
    question: str
    concept_id: Optional[str] = None
    options: List[str] = Field(default_factory=list)
    allow_freeform: bool = True


class CheckpointResponsePayload(BaseModel):
    """Student responded to a checkpoint."""

    checkpoint_id: str
    response: str
    is_correct: Optional[bool] = None
    score: Optional[float] = None
    feedback: Optional[str] = None


class SourceCitationPayload(BaseModel):
    """Source reference attached to tutor content."""

    citation_id: str
    resource_id: str
    resource_name: Optional[str] = None
    chunk_ids: List[str] = Field(default_factory=list)
    page_numbers: List[int] = Field(default_factory=list)
    snippet: Optional[str] = None


class WarningPayload(BaseModel):
    """Operational warning surfaced to UI."""

    warning_type: str
    message: str
    selected_model_id: Optional[str] = None
    routed_model_id: Optional[str] = None
    reroute_reason: Optional[str] = None


class SessionCompletedPayload(BaseModel):
    """Session completion summary."""

    session_id: UUID
    summary_text: Optional[str] = None
    concepts_strong: List[str] = Field(default_factory=list)
    concepts_developing: List[str] = Field(default_factory=list)
    concepts_to_revisit: List[str] = Field(default_factory=list)
    objectives: List[ObjectiveSnapshot] = Field(default_factory=list)
    mastery_snapshot: Dict[str, float] = Field(default_factory=dict)
    turn_count: int = 0
    recommended_next: Optional[str] = None


# ---------------------------------------------------------------------------
# Envelope
# ---------------------------------------------------------------------------


class SessionEvent(BaseModel):
    """Canonical event envelope for the live study workspace."""

    event_type: SessionEventType
    timestamp: datetime = Field(default_factory=lambda: datetime.now())
    session_id: UUID
    notebook_id: Optional[UUID] = None
    payload: Dict[str, Any] = Field(default_factory=dict)
    is_persistent: bool = False  # True = stored in DB, False = transient


# ---------------------------------------------------------------------------
# Helper to construct events
# ---------------------------------------------------------------------------


def make_event(
    event_type: SessionEventType,
    session_id: UUID,
    payload: BaseModel | Dict[str, Any],
    *,
    notebook_id: UUID | None = None,
    is_persistent: bool = False,
) -> SessionEvent:
    payload_dict = payload.model_dump() if isinstance(payload, BaseModel) else payload
    return SessionEvent(
        event_type=event_type,
        session_id=session_id,
        notebook_id=notebook_id,
        payload=payload_dict,
        is_persistent=is_persistent,
    )
